// baojs/dist/router/router.js
import { Node } from "./tree";
const httpMethods = [
    "GET",
    "HEAD",
    "POST",
    "PUT",
    "DELETE",
    "CONNECT",
    "OPTIONS",
    "TRACE",
    "PATCH",
];
const NOT_FOUND = { handler: null, params: {} };
export class BaoRouter {
    trees;
    opts;
    constructor(opts = {}) {
        if (!(this instanceof BaoRouter)) {
            return new BaoRouter(opts);
        }
        this.trees = {};
        this.opts = opts;
    }
    any(path, handler) {
        for (const method of httpMethods) {
            this.on(method, path, handler);
        }
    }
    ws(path, handlers) {
        if (path[0] !== "/") {
            throw new Error("path must begin with '/' in path");
        }
        if (!this.trees["WS"]) {
            this.trees["WS"] = new Node();
        }
        this.trees["WS"].addRoute(path, handlers);
        return this;
    }
    on(method, path, handler) {
        if (path[0] !== "/") {
            throw new Error("path must begin with '/' in path");
        }
        if (!this.trees[method]) {
            this.trees[method] = new Node();
        }
        this.trees[method].addRoute(path, handler);
        return this;
    }
    find(method, path) {
        const tree = this.trees[method];
        if (tree) {
            return tree.search(path);
        }
        return NOT_FOUND;
    }
}
